### ISSUE TITLE
indirectly-imported recursive types causes type-identity confusion

### ISSUE THREAD

**Issue by bukzor** [microsoft/pyright#8784]
I'm getting a false `"(type X)" is not the same as "(*exactly the same* type X)"` error in the below, very specific, situation.

If I change anything about this demo (remove the seemingly unrelated function, change the order, remove the JSONArray alternate from the typealias, ...) then the error goes away, but I encountered this while programming a real system.

```console
$ tail *
==> metric_gocd_raw.py <==
import typedefs
def func2(_: typedefs.JSONObject) -> None: pass
def main(x: typedefs.JSON) -> None:
    assert isinstance(x, list), x
    y: list[typedefs.JSON] = x
    print(y)

==> pyproject.toml <==

==> typedefs.py <==
JSONArray = list["JSON"]
JSONObject = dict[str, "JSON"]
JSON = JSONObject | JSONArray

$ pyright --verbose . | ~/bin/symbolic-paths
Loading pyproject.toml file at $PWD/pyproject.toml
Pyproject file "$PWD/pyproject.toml" has no "[tool.pyright]" section.
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
Search paths for file://$PWD
  $XDG_CACHE_HOME/pyright-python/1.1.376/node_modules/pyright/dist/typeshed-fallback/stdlib
  $PWD
  $PWD/typings
  $XDG_CACHE_HOME/pyright-python/1.1.376/node_modules/pyright/dist/typeshed-fallback/stubs/...
  $VIRTUAL_ENV_BASE_PREFIX/lib/python3.12
  $VIRTUAL_ENV_BASE_PREFIX/lib/python3.12/lib-dynload
  $VIRTUAL_ENV/lib/python3.12/site-packages
Found 2 source files
pyright 1.1.376
$PWD/metric_gocd_raw.py
  $PWD/metric_gocd_raw.py:5:30 - error: Expression of type "builtins.list[JSON]" is incompatible with declared type "builtins.list[typedefs.JSON]"
    "builtins.list" is incompatible with "builtins.list"
      Type parameter "_T@list" is invariant, but "JSON" is not the same as "typedefs.JSON"
      Consider switching from "list" to "Sequence" which is covariant (reportAssignmentType)
1 error, 0 warnings, 0 informations
```

**Comment by bukzor** [comment]
Just inferring from the observed behavior, and knowing quite little about pyright internals, it looks like:

1. the `JSON` type is constructed by the reference to `JSONObject` at `func2` (lazily reified?)
2. a *different* `JSON` type is constructed by the reference to `typedefs.JSONObject` at `main`
3. the two instances have different names (#1 is named `JSON` while #2 is named `typedefs.JSON`) and are not considered invariantly equal, even though they in fact stem from the *exact* same definition

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.377.
